### PR TITLE
Support OpenMP Reduction

### PR DESCRIPTION
**Pull request by BradSwain**
## Old Problems

There are two problems with the ~~current~~ old solution.

1. We use calls to `reduce` and `reduce_end` to mark the reduction code, but this is not correct.

There can be cases like
```
%15 = call i32 @__kmpc_reduce(...)
  switch i32 %15, label %.omp.reduction.default [
    i32 1, label %.omp.reduction.case1
    i32 2, label %.omp.reduction.case2
  ], !dbg !52

.omp.reduction.case1:
  ...
  call void @__kmpc_end_reduce(...)
  br label %.omp.reduction.default

.omp.reduction.case2:
  call void @__kmpc_end_reduce(...)
  br label %.omp.reduction.default, !dbg !52
```
where there are two calls to `end_reduce`

2. We originally assumed reduction code cannot be part of a race.

According to the OpenMP Specification and examples, this can be a race
```
#pragma omp parallel
{
  #pragma omp master
  shared  = 1;
  // Potential race on shared unless barrier here
  #pragma omp for reduction(+:shared)
  for (...) { ... }
```

### Problem 1. Identifying code in reduce region

Don't rely on openmp reduce start/end to identify blocks belonging to reduce.

From the example above:
```cpp
%15 = call i32 @__kmpc_reduce(...)
  switch i32 %15, label %.omp.reduction.default [
    i32 1, label %.omp.reduction.case1
    i32 2, label %.omp.reduction.case2
  ], !dbg !52

.omp.reduction.case1:
  ...
  call void @__kmpc_end_reduce(...)
  br label %.omp.reduction.default

.omp.reduction.case2:
  call void @__kmpc_end_reduce(...)
  br label %.omp.reduction.default, !dbg !52
```

now when we see a call to `__kmpc_reduce`, we look at the next instruction which should always:
 1. be a switch
 2. have a default case jumping to the end of the reduce region

Now we check to see which blocks are "between" the switch and the block that is jumped to by default.

Between means, we find all blocks reachable from switch, but stop at the `%.omp.reduction.default` block.

### Problem 2. When can reduce code race

For now we treat reduce and reduce_nowait the same. 

We treat them as if they cannot race with themselves, but may race with anything else.

### CODE REVIEW

**Comment by BradSwain** [comment]
Closes #94

File: src/RaceDetect.cpp [hunk 1]
```diff
@@ -49,13 +49,24 @@ Report race::detectRaces(llvm::Module *module, DetectRaceConfig config) {
...
       }
 
-      if (ompAnalysis.inSameSingleBlock(write, other)) {
+      // Certain omp blocks cannot race with themselves
+      if (ompAnalysis.inSameSingleBlock(write, other) || ompAnalysis.inSameReduce(write, other)) {
         return;
       }
+
+      // We assume any event inside of a non-nowait version of reduce can never be part of a race
+      // if (ompAnalysis.inReduce(write) || ompAnalysis.inReduce(other)) {
```
Comment (funemy) [inline src/RaceDetect.cpp:67]: remove this piece of dead code

File: src/Analysis/OpenMPAnalysis.cpp [hunk 2]
```diff
@@ -45,6 +46,26 @@ bool OpenMPAnalysis::canIndexOverlap(const race::MemAccessEvent* event1, const r
 }
 
 namespace {
+
+// return true if both events belong to the same OpenMP team
+bool _inSameTeam(const Event* event1, const Event* event2) {
```
Comment (funemy) [inline src/Analysis/OpenMPAnalysis.cpp:51]: Maybe add a comment clarify why this function need to be split out (in this case it's for the assertion)

File: src/Analysis/OpenMPAnalysis.cpp [hunk 3]
```diff
@@ -74,26 +95,68 @@ std::vector<Region> getRegions(const ThreadTrace& thread) {
...
+  }
+  return false;
+}
+
+auto const _inSameSingleBlock = inSame<IR::Type::OpenMPSingleStart, IR::Type::OpenMPSingleEnd>;
+
 }  // namespace
 
-const std::vector<OpenMPAnalysis::LoopRegion>& OpenMPAnalysis::getOmpForLoops(const ThreadTrace& thread) {
+const std::vector<OpenMPAnalysis::LoopRegion>& OpenMPAnalysis::getOmpForLoopsCached(const ThreadTrace& thread) {
```
Comment (funemy) [inline src/Analysis/OpenMPAnalysis.cpp:144]: If this function name doesn't conflict with others, change it back?

File: src/Analysis/OpenMPAnalysis.cpp [hunk 4]
```diff
@@ -114,43 +177,92 @@ bool OpenMPAnalysis::isLoopArrayAccess(const race::MemAccessEvent* event1, const
...
-  auto const eid2 = e2Spawn.value()->getID();
-  auto const diff = (eid1 > eid2) ? (eid1 - eid2) : (eid2 - eid1);
-  return diff == 1;
-}
+  // compute results, cache them, then return them
+  auto& blocks = reduceBlocks[reduce];
 
-bool OpenMPAnalysis::inSameSingleBlock(const Event* event1, const Event* event2) const {
-  assert(inSameTeam(event1, event2));
+  auto const switchInst = llvm::dyn_cast<llvm::SwitchInst>(reduce->getNextNode());
```
Comment (funemy) [inline src/Analysis/OpenMPAnalysis.cpp:193]: Add a comment explaining the pattern of reduce and its corresponding switch

**Review by funemy — APPROVED** [review]

**Pull request merged**